### PR TITLE
[OTA-R] OTA watchdog timer should not expire when device is in idle state.

### LINKED ISSUE #18222: OTA watchdog timer should not expire when device is in idle state.
#### Problem

OTA watchdog timer should not expire when device is in idle state.  

```
[1651703424056] [22377:597089] CHIP: [SWU] Stopping the watchdog timer
[1651703424056] [22377:597089] CHIP: [SWU] Starting the periodic query timer, timeout: 86400 seconds
[1651703424059] [22377:597089] CHIP: [DMG] All ReadHandler-s are clean, clear GlobalDirtySet
[1651703424060] [22377:597089] CHIP: [SWU] OTA image downloaded to /tmp/test.bin
[1651703424258] [22377:597089] CHIP: [EM] Received message of type 0x10 with protocolId (0, 0) and MessageCounter:10947771 on exchange 39081i
[1651703424258] [22377:597089] CHIP: [EM] Found matching exchange: 39081i, Delegate: 0x0
[1651703424258] [22377:597089] CHIP: [EM] Rxd Ack; Removing MessageCounter:7473674 from Retrans Table on exchange 39081i
[1651703424258] [22377:597089] CHIP: [EM] Removed CHIP MessageCounter:7473674 from RetransTable on exchange 39081i
[1651704294908] [22377:607983] CHIP: [DIS] OnExpiration callback for cleared session
[1651704294908] [22377:607983] CHIP: [DIS] OnDiscoveryExpiration callback for cleared session
*[1651726325897] [22377:760473] CHIP: [SWU] Watchdog timer detects state stuck at 1. Cancelling download and resetting state.*
*[1651726325899] [22377:760473] CHIP: [BDX] No download in progress*
*[1651726325899] [22377:760473] CHIP: [ZCL] Previous state and new state are the same (1), no event to log*
[1651726326268] [22377:760473] CHIP: [SWU] Starting the periodic query timer, timeout: 86400 seconds
```

In `DefaultOTARequestorDriver::StopPeriodicQueryTimer()` and `DefaultOTARequestorDriver::StopWatchdogTimer()`, need to update the `CancelDelayedAction()` call to pass in the correct `TimerCompleteCallback` for the timer to be cancelled.  Upon further debugging of this issue, found that both watchdog and periodic timers are not being cancelled when their respective `StopWatchdogTimer()` and `StopPeriodicQueryTimer()` functions are being called.  This is due to `TimerCompleteCallback aOnComplete` mismatch in the `TimerList::Remove()` call.

`TimerList::Node * TimerList::Remove(TimerCompleteCallback aOnComplete, void * aAppState)`

 #### Proposed Solution

Ensure OTA watchdog timer is only running in idle state and periodic query timer is running in non-idle state by fixing the bug in `DefaultOTARequestorDriver::StopPeriodicQueryTimer()` and `DefaultOTARequestorDriver::StopWatchdogTimer()`.

### PR DESCRIPTION
**Pull request by isiu-apple**
#### Problem

OTA watchdog timer should not expire when device is in idle state.  

```
[1651703424056] [22377:597089] CHIP: [SWU] Stopping the watchdog timer
[1651703424056] [22377:597089] CHIP: [SWU] Starting the periodic query timer, timeout: 86400 seconds
[1651703424059] [22377:597089] CHIP: [DMG] All ReadHandler-s are clean, clear GlobalDirtySet
[1651703424060] [22377:597089] CHIP: [SWU] OTA image downloaded to /tmp/test.bin
[1651703424258] [22377:597089] CHIP: [EM] Received message of type 0x10 with protocolId (0, 0) and MessageCounter:10947771 on exchange 39081i
[1651703424258] [22377:597089] CHIP: [EM] Found matching exchange: 39081i, Delegate: 0x0
[1651703424258] [22377:597089] CHIP: [EM] Rxd Ack; Removing MessageCounter:7473674 from Retrans Table on exchange 39081i
[1651703424258] [22377:597089] CHIP: [EM] Removed CHIP MessageCounter:7473674 from RetransTable on exchange 39081i
[1651704294908] [22377:607983] CHIP: [DIS] OnExpiration callback for cleared session
[1651704294908] [22377:607983] CHIP: [DIS] OnDiscoveryExpiration callback for cleared session
*[1651726325897] [22377:760473] CHIP: [SWU] Watchdog timer detects state stuck at 1. Cancelling download and resetting state.*
*[1651726325899] [22377:760473] CHIP: [BDX] No download in progress*
*[1651726325899] [22377:760473] CHIP: [ZCL] Previous state and new state are the same (1), no event to log*
[1651726326268] [22377:760473] CHIP: [SWU] Starting the periodic query timer, timeout: 86400 seconds
```

In `DefaultOTARequestorDriver::StopPeriodicQueryTimer()` and `DefaultOTARequestorDriver::StopWatchdogTimer()`, need to update the `CancelDelayedAction()` call to pass in the correct `TimerCompleteCallback` for the timer to be cancelled.  Upon further debugging of this issue, found that both watchdog and periodic timers are not being cancelled when their respective `StopWatchdogTimer()` and `StopPeriodicQueryTimer()` functions are being called.  This is due to `TimerCompleteCallback aOnComplete` mismatch in the `TimerList::Remove()` call.

`TimerList::Node * TimerList::Remove(TimerCompleteCallback aOnComplete, void * aAppState)`

Fixes: https://github.com/project-chip/connectedhomeip/issues/18222

#### Change overview

Pass in correct watchdog and periodic timer handler functions as `TimerCompleteCallback` when starting and cancelling the respective timers.

#### Testing

- Set breakpoints in `DefaultOTARequestorDriver::StopWatchdogTimer()` and `DefaultOTARequestorDriver::StopPeriodicQueryTimer()` and ensured the timers were properly removed in `TimerList::Remove()`.
- Verified Query image works.
- Verified watchdog timer and periodic timer works.
- Verified watchdog timer and periodic timer are no longer running at the same time.

### CODE REVIEW

File: src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp [hunk 1]
```diff
@@ -319,60 +319,61 @@ void DefaultOTARequestorDriver::PeriodicQueryTimerHandler(System::Layer * system
 {
     ChipLogProgress(SoftwareUpdate, "Default Provider timer handler is invoked");
 
+    DefaultOTARequestorDriver * driver = reinterpret_cast<DefaultOTARequestorDriver *>(appState);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp:322]:
```suggestion
    DefaultOTARequestorDriver * driver = ToDriver(appState);
```

File: src/app/clusters/ota-requestor/DefaultOTARequestorDriver.h [hunk 2]
```diff
@@ -84,10 +84,10 @@ class DefaultOTARequestorDriver : public OTARequestorDriver
 protected:
     void StartPeriodicQueryTimer();
     void StopPeriodicQueryTimer();
-    void PeriodicQueryTimerHandler(System::Layer * systemLayer, void * appState);
+    static void PeriodicQueryTimerHandler(System::Layer * systemLayer, void * appState);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/DefaultOTARequestorDriver.h:87]: nit: Maybe group the static functions at the top to separate from non-static functions for easier reading.

File: src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp [hunk 3]
```diff
@@ -319,60 +319,61 @@ void DefaultOTARequestorDriver::PeriodicQueryTimerHandler(System::Layer * system
...
 {
     ChipLogProgress(SoftwareUpdate, "Stopping the Periodic Query timer");
-    CancelDelayedAction([](System::Layer *, void * context) { (ToDriver(context))->PeriodicQueryTimerHandler(nullptr, context); },
-                        this);
+    CancelDelayedAction(PeriodicQueryTimerHandler, this);
 }
 
 void DefaultOTARequestorDriver::WatchdogTimerHandler(System::Layer * systemLayer, void * appState)
 {
+    DefaultOTARequestorDriver * driver = reinterpret_cast<DefaultOTARequestorDriver *>(appState);
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp:354]:
```suggestion
    DefaultOTARequestorDriver * driver = ToDriver(appState);
```

**Review by carol-apple — APPROVED** [review]

**Review by selissia — APPROVED** [review]

File: src/app/clusters/ota-requestor/DefaultOTARequestorDriver.h (hunk 2 above)
Comment (isiu-apple) [reply src/app/clusters/ota-requestor/DefaultOTARequestorDriver.h]: Done.

File: src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp (hunk 1 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/DefaultOTARequestorDriver.cpp:322]: Wait, I don't think the change is made?

**Pull request merged**